File: android/src/main/kotlin/com/microsoft/clarity/ClarityFlutterPlugin.kt
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

package com.microsoft.clarity

import android.content.Context
import com.microsoft.clarity.api.ClarityDeviceHostApiImpl
import com.microsoft.clarity.api.ClarityGaidHostApiImpl
import com.microsoft.clarity.api.ClarityNetworkHostApiImpl
import com.microsoft.clarity.api.ConnectivityChangedStreamHandlerImpl
import com.microsoft.clarity.bridges.NetworkBridge
import com.microsoft.clarity.generated.ClarityDeviceHostApi
import com.microsoft.clarity.generated.ClarityGaidHostApi
import com.microsoft.clarity.generated.ClarityNetworkHostApi
import io.flutter.embedding.engine.plugins.FlutterPlugin

class ClarityFlutterPlugin : FlutterPlugin {
    private var networkBridge: NetworkBridge? = null
    private var connectivityStreamHandler: ConnectivityChangedStreamHandlerImpl? = null

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        val context: Context = binding.applicationContext
        val messenger = binding.binaryMessenger

        ClarityDeviceHostApi.setUp(messenger, ClarityDeviceHostApiImpl(context))
        ClarityGaidHostApi.setUp(messenger, ClarityGaidHostApiImpl(context))

        val bridge = NetworkBridge(context)
        networkBridge = bridge
        ClarityNetworkHostApi.setUp(messenger, ClarityNetworkHostApiImpl(bridge))
        connectivityStreamHandler =
            ConnectivityChangedStreamHandlerImpl.register(messenger, bridge)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        ClarityDeviceHostApi.setUp(binding.binaryMessenger, null)
        ClarityGaidHostApi.setUp(binding.binaryMessenger, null)
        ClarityNetworkHostApi.setUp(binding.binaryMessenger, null)
        connectivityStreamHandler?.dispose()
        connectivityStreamHandler = null
        networkBridge = null
    }
}

File: ios/clarity_flutter/Sources/clarity_flutter/ClarityFlutterPlugin.swift
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

import Flutter

public class ClarityFlutterPlugin: NSObject, FlutterPlugin {

    private var networkBridge: NetworkBridge?
    private var connectivityStreamHandler: ConnectivityChangedStreamHandlerImpl?

    public static func register(with registrar: FlutterPluginRegistrar) {
        let instance = ClarityFlutterPlugin()
        let messenger = registrar.messenger()

        ClarityDeviceHostApiSetup.setUp(
            binaryMessenger: messenger,
            api: ClarityDeviceHostApiImpl()
        )

        ClarityGaidHostApiSetup.setUp(
            binaryMessenger: messenger,
            api: ClarityGaidHostApiImpl()
        )

        let bridge = NetworkBridge()
        instance.networkBridge = bridge
        ClarityNetworkHostApiSetup.setUp(
            binaryMessenger: messenger,
            api: ClarityNetworkHostApiImpl(networkBridge: bridge)
        )
        let streamHandler = ConnectivityChangedStreamHandlerImpl(networkBridge: bridge)
        instance.connectivityStreamHandler = streamHandler
        ConnectivityChangedStreamHandler.register(
            with: messenger,
            streamHandler: streamHandler
        )
    }

    public func detachFromEngine(for registrar: FlutterPluginRegistrar) {
        let messenger = registrar.messenger()
        ClarityDeviceHostApiSetup.setUp(binaryMessenger: messenger, api: nil)
        ClarityGaidHostApiSetup.setUp(binaryMessenger: messenger, api: nil)
        ClarityNetworkHostApiSetup.setUp(binaryMessenger: messenger, api: nil)
        connectivityStreamHandler?.dispose()
        connectivityStreamHandler = nil
        networkBridge = nil
    }
}
